src/Shared-Src/Datadog.Util/internal/ExceptionAggregator.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Util
+{
+    internal ref struct ExceptionAggregator

```

Comment:
**Why struct:**

The optimization here is to save on heap allocations. When this type does:
if no exception thrown => no allocations. That was the main goal since exception already means slow path. But even then:
One exception: only an additional allocation when re-throwing (i.e. a slow operation already happening) - the Dispatch Info.
Multiple exceptions: well, we are in a slow mode just due to them already. We get away with two allocations (see comment).

If we made this a class, we would always have an allocation - just for creating an instance on the aggregator, even if there are no exceptions. So `ref` does not make things faster, it just prevents a potential issue with boxing. It's the `struct` that makes things faster by preventing an allocation.

**Why not immutable:**
Well, the type is conceptually mutable: it has an object fied that can point to an exception of a list of exceptions (or be null). If you boxed such a struct and then modified the boxed copy, you would modify the pointer inside the boxed copy, leaving the pointer in the original struct unaffected. This is the common problem with mutable value types. It's an interesting topic, we can discuss on Zoom :)

PR was approved. Merging and tagging @gleocadie to take note of this comment. :)